c/parallel/src/radix_sort.cu

```diff
@@ -0,0 +1,843 @@
+
+  int BlockThreads() const
+  {
+    return block_threads;
+  }
+
+  int ItemsPerThread() const
+  {
+    return items_per_thread;
+  }
+};
+
+struct agent_radix_sort_onesweep_policy
+{
+  int block_threads;
+  int items_per_thread;
+  int rank_num_parts;
+  int radix_bits;
+
+  int BlockThreads() const
+  {
+    return block_threads;
+  }
+
+  int ItemsPerThread() const
+  {
+    return items_per_thread;
+  }
+};
+
+struct agent_radix_sort_histogram_policy
+{
+  int block_threads;
+  int items_per_thread;
+  int num_parts;
+  int radix_bits;
+
+  int BlockThreads() const
+  {
+    return block_threads;
+  }
+};
+
+struct agent_radix_sort_exclusive_sum_policy
+{
+  int block_threads;
+  int radix_bits;
+};
+
+struct agent_scan_policy
+{
+  int block_threads;
+  int items_per_thread;
+
+  int BlockThreads() const
+  {
+    return block_threads;
+  }
+
+  int ItemsPerThread() const
+  {
+    return items_per_thread;
+  }
+};
+
+struct radix_sort_runtime_tuning_policy
+{
+  agent_radix_sort_histogram_policy histogram;
+  agent_radix_sort_exclusive_sum_policy exclusive_sum;
+  agent_radix_sort_onesweep_policy onesweep;
+  agent_scan_policy scan;
+  agent_radix_sort_downsweep_policy downsweep;
+  agent_radix_sort_downsweep_policy alt_downsweep;
+  agent_radix_sort_upsweep_policy upsweep;
+  agent_radix_sort_upsweep_policy alt_upsweep;
+  agent_radix_sort_downsweep_policy single_tile;
+  bool is_onesweep;
+
+  agent_radix_sort_histogram_policy Histogram() const
+  {
+    return histogram;
+  }
+
+  agent_radix_sort_exclusive_sum_policy ExclusiveSum() const
+  {
+    return exclusive_sum;
+  }
+
+  agent_radix_sort_onesweep_policy Onesweep() const
+  {
+    return onesweep;
+  }
+
+  agent_scan_policy Scan() const
+  {
+    return scan;
+  }
+
+  agent_radix_sort_downsweep_policy Downsweep() const
+  {
+    return downsweep;
+  }
+
+  agent_radix_sort_downsweep_policy AltDownsweep() const
+  {
+    return alt_downsweep;
+  }
+
+  agent_radix_sort_upsweep_policy Upsweep() const
+  {
+    return upsweep;
+  }
+
+  agent_radix_sort_upsweep_policy AltUpsweep() const
+  {
+    return alt_upsweep;
+  }
+
+  agent_radix_sort_downsweep_policy SingleTile() const
+  {
+    return single_tile;
+  }
+
+  bool IsOnesweep() const
+  {
+    return is_onesweep;
+  }
+
+  template <typename PolicyT>
+  CUB_RUNTIME_FUNCTION static constexpr int RadixBits(PolicyT policy)
+  {
+    return policy.radix_bits;
+  }
+
+  template <typename PolicyT>
+  CUB_RUNTIME_FUNCTION static constexpr int BlockThreads(PolicyT policy)
+  {
+    return policy.block_threads;
+  }
+};
+
+std::pair<int, int>
+reg_bound_scaling(int nominal_4_byte_block_threads, int nominal_4_byte_items_per_thread, int key_size)
+{
+  int items_per_thread = std::max(1, nominal_4_byte_items_per_thread * 4 / std::max(4, key_size));
+  int block_threads =
+    std::min(nominal_4_byte_block_threads,
+             cuda::ceil_div(int{cub::detail::max_smem_per_block} / (key_size * items_per_thread), 32) * 32);
```

Comment:
I think we should handle the case `key_size == 0`, or add assertions that each input parameter is strictly positive.